jsonrpc/server.go

```diff
@@ -0,0 +1,279 @@
+				Version: "2.0",
+				Error:   err,
+			}
+		} else {
+			res = s.handleRequest(req)
+		}
+
+		return json.Marshal(res)
+	} else {
+		batchReq := batch{}
+		batchRes := []json.RawMessage{}
+
+		if err := json.NewDecoder(bufferedReader).Decode(&batchReq); err != nil {
+			return nil, err
+		}
+
+		for _, rawReq := range batchReq {
+			if res, err := s.Handle(rawReq); err == nil { // todo: handle async
+				batchRes = append(batchRes, res)
+			}
+		}
+
+		if len(batchRes) == 0 {
+			return nil, nil
+		}
+		return json.Marshal(batchRes)
+	}
+}
+
+func isBatch(reader *bufio.Reader) bool {
+	for {
+		char, err := reader.ReadByte()
+		if err != nil {
+			break
+		} else if char == ' ' || char == '\t' || char == '\r' || char == '\n' {
+			continue
+		} else {
+			reader.UnreadByte()
+			return char == '['
+		}
+	}
+
+	return false
+}
+
+func isNil(i any) bool {
+	return i == nil || reflect.ValueOf(i).IsNil()
+}
+
+func (s *Server) handleRequest(req *request) *response {
+	res := &response{
+		Version: "2.0",
+		Id:      req.Id,
+	}
+
+	calledMethod, found := s.methods[req.Method]
+	if !found {
+		res.Error = &Error{
+			Code:    MethodNotFound,
+			Message: "method not found",
+		}
+		return res
+	}
+
+	args, err := buildArguments(req.Params, calledMethod.Handler, calledMethod.Params)
+	if err != nil {
+		res.Error = &Error{
+			Code:    InvalidParams,
+			Message: err.Error(),
+		}
+		return res
+	}
+
+	tuple := reflect.ValueOf(calledMethod.Handler).Call(args)
+	if errAny := tuple[1].Interface(); !isNil(errAny) {
+		res.Error = errAny.(*Error)
+		return res
+	}
+
+	res.Result = tuple[0].Interface()
+	return res
+}
+
+func buildArguments(params, handler any, configuredParams []Parameter) ([]reflect.Value, error) {
+	args := []reflect.Value{}
+	handlerType := reflect.TypeOf(handler)
+	paramCount := handlerType.NumIn()
+	paramsKind := reflect.TypeOf(params).Kind()
+
+	for idx := 0; idx < paramCount; idx++ {
+		valueContainer := reflect.New(handlerType.In(idx))
+		var requestValue any
+		found := true
+
+		switch paramsKind {
+		case reflect.Slice:
+			paramsList := params.([]any)
+			if len(paramsList) != paramCount {
+				return nil, errors.New("missing param in list")
+			}
+
+			requestValue = paramsList[idx]
+		case reflect.Map:
+			paramsMap := params.(map[string]any)
+			if len(configuredParams) != paramCount {
+				return nil, errors.New("missing param name")
+			}
+
+			requestValue, found = paramsMap[configuredParams[idx].Name]
+		default:
+			return nil, errors.New("impossible param type: check request.isSane")
+		}
+
+		if found {
+			valueMarshaled, err := json.Marshal(requestValue)
+			if err != nil {
+				return nil, err
+			}
+			err = json.Unmarshal(valueMarshaled, valueContainer.Interface())
+			if err != nil {
+				return nil, err
+			}
+		} else if !configuredParams[idx].Optional {
+			return nil, errors.New("missing non-optional param")
+		}
+
+		args = append(args, valueContainer.Elem())
+	}
+
+	return args, nil
+}
```

Comment:
Some comments would be helpful. It took me a while to understand what was happening here.